### PR TITLE
sanitize data coming from upload-page

### PR DESCRIPTION
**Pull request by martinobersteiner**
Data that users put into the upload-page was barely sanitized.

This PR adds sanitization.

### CODE REVIEW

File: invenio_records_lom/services/schemas/metadata.py [hunk 1]
```diff
@@ -56,27 +74,77 @@ def _deserialize(self, value, attr, data, **kwargs):
 
         # validate inner keys
         if not isinstance(langstring_inner, dict):
-            raise self.make_error("invalid")
+            raise self.make_error("invalid_inner_type")
         inner_keys = set(langstring_inner)
         if "#text" not in inner_keys:
             raise self.make_error("missing_langstring_key", keys="#text")
-        if "lang" not in inner_keys and self.lang_exists:
+        if "lang" not in inner_keys and self.lang_exists is True:
```
Comment (utnapischtim) [inline invenio_records_lom/services/schemas/metadata.py:81]: minor: why do you add the explicit check to True?

File: invenio_records_lom/services/schemas/metadata.py [hunk 2]
```diff
@@ -56,27 +74,77 @@ def _deserialize(self, value, attr, data, **kwargs):
...
+        if "lang" in inner_keys and self.lang_exists is False:
             raise self.make_error("extraneous_lang")
         if not inner_keys <= {"#text", "lang"}:
             raise self.make_error(
                 "extraneous_keys", keys=sorted(inner_keys - {"#text", "lang"})
             )
 
-        text = langstring_inner["#text"]
-        lang = langstring_inner.get("lang", True)
+        text = sanitize_unicode(str(langstring_inner["#text"]))
```
Comment (utnapischtim) [inline invenio_records_lom/services/schemas/metadata.py:90]: minor: why is it necessary to cast to str?

File: invenio_records_lom/services/schemas/metadata.py [hunk 3]
```diff
@@ -56,27 +74,77 @@ def _deserialize(self, value, attr, data, **kwargs):
...
+
+        return result
+
+
+def validate_langstring_lang(validator_callable):
+    """Wraps `validator_callable`, passing it value["langstring"]["lang"] instead of value itself."""
+
+    def validate_lang(value):
+        try:
+            value = value["langstring"]["lang"]
```
Comment (utnapischtim) [inline invenio_records_lom/services/schemas/metadata.py:113]: style: don't change type of variable. `value` was as the parameter a `dict`, after this assignment it is a `str`. i know it is possible in python, but please don't do it.

i think `lang` would be the better name for the variable. this helps also to see that the value given to `validator_callable` is not the same as the `validate_lang` function received.

File: invenio_records_lom/services/schemas/metadata.py [hunk 4]
```diff
@@ -56,27 +74,77 @@ def _deserialize(self, value, attr, data, **kwargs):
...
+
+    return validate_lang
+
+
+def validate_langstring_text(validator_callable):
+    """Wraps `validator_callable`, passing it value["langstring"]["#text"] instead of value itself."""
+
+    def validate_text(value):
+        try:
+            value = value["langstring"]["#text"]
```
Comment (utnapischtim) [inline invenio_records_lom/services/schemas/metadata.py:126]: style: see above

File: invenio_records_lom/services/schemas/metadata.py [hunk 5]
```diff
@@ -105,24 +216,56 @@ class LifecycleSchema(Schema):
...
     contribute = fields.List(
         fields.Nested(ContributeSchema()),
-        validate=validate.Length(min=1, error="Enter at least one contribution."),
+        required=True,
     )
 
+    @validates("contribute")
+    def validate_author_publisher_exist(self, contributes):
+        """Validates existence of >=1 author/publisher contributes."""
+        author_contributes = [
```
Comment (utnapischtim) [inline invenio_records_lom/services/schemas/metadata.py:225]: style: could you please rewrite this. this is hard to read.

suggestion:
```
def get_lower_text(contribute):
    try:
       return contribute["role"]["value"]["langstring"]["#text"].lower()
    except KeyError:
       return ""
author_contributes = [contribute for contribute in contributes if get_lower_text(contribute) == "author"]
```

File: invenio_records_lom/services/schemas/metadata.py (hunk 1 above)
Comment (martinobersteiner) [reply invenio_records_lom/services/schemas/metadata.py:81]: `self.lang_exists` needs to distinguish *three* (not two) cases:
1. validate `lang` definitely exists (`lang_exists is True`)
2. validate `lang` definitely does *not* exist (`lang_exists is False`)
3. don't validate either way (`lang_exists is None`)

for three cases truthy vs falsy isn't enough...
to reflect its three-way nature, I thought about renaming it, but couldn't come up with a better name...
feel free to suggest one!

File: invenio_records_lom/services/schemas/metadata.py (hunk 2 above)
Comment (martinobersteiner) [reply invenio_records_lom/services/schemas/metadata.py:90]: not technically necessary
in my manual testing of schemas, where I can actually assign any value to `#text`, the error-message for assigning an int was hard to read
I don't think there even is a way for it to be not a `str` at this point...
can remove this if you want?

File: invenio_records_lom/services/schemas/metadata.py (hunk 3 above)
Comment (martinobersteiner) [reply invenio_records_lom/services/schemas/metadata.py:113]: Done.

File: invenio_records_lom/services/schemas/metadata.py (hunk 4 above)
Comment (martinobersteiner) [reply invenio_records_lom/services/schemas/metadata.py:126]: see above

File: invenio_records_lom/services/schemas/metadata.py (hunk 1 above)
Comment (utnapischtim) [reply invenio_records_lom/services/schemas/metadata.py:81]: ohhh, i see. yeah a new name is necessary. for me a variable with an prefix/postfix of `exists` is a bool value with only True/False values and None is False by definition.

where does point 3 jumps in? where is this calculated?

**Review by utnapischtim — APPROVED** [review]

File: invenio_records_lom/services/schemas/metadata.py (hunk 1 above)
Comment (martinobersteiner) [reply invenio_records_lom/services/schemas/metadata.py]: resolved by using two booleans instead of a weird three-way variable

**Pull request merged**